File: sdks/python/cvt_sdk/producer/testing.py
"""
Producer testing module for CVT.

Provides ProducerTestKit for schema compliance testing - allows producers
to validate their API responses against their OpenAPI schema without
needing real consumers.

Example with TLS and API key:
    >>> from cvt_sdk.producer import ProducerTestKit, ProducerTestConfig, TLSOptions
    >>>
    >>> # Create test kit
    >>> test_kit = ProducerTestKit(ProducerTestConfig(
    ...     schema_id="user-api",
    ...     server_address="localhost:50051",
    ...     tls=TLSOptions(enabled=True, root_cert_path="./certs/ca.crt"),
    ...     api_key="cvt-dev-key-12345",
    ... ))
    >>>
    >>> # Validate a response
    >>> result = test_kit.validate_response(
    ...     method="GET",
    ...     path="/users/123",
    ...     status_code=200,
    ...     body={"id": "123", "name": "John Doe"},
    ... )
    >>>
    >>> assert result.valid, f"Validation failed: {result.errors}"
"""

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional

import grpc

from ..proto import (
    ContractValidatorStub,
    RequestData,
    ResponseData,
    ValidateProducerRequest,
    ValidationResult as ProtoValidationResult,
)


@dataclass
class TLSOptions:
    """TLS configuration options for secure connections."""

    __test__ = False  # Prevent pytest from collecting this class

    enabled: bool = False
    """Enable TLS for the connection."""

    root_cert_path: Optional[str] = None
    """Path to the root CA certificate file for server verification."""

    cert_path: Optional[str] = None
    """Path to the client certificate file (for mTLS)."""

    key_path: Optional[str] = None
    """Path to the client private key file (for mTLS)."""


@dataclass
class ProducerTestConfig:
    """Configuration for the ProducerTestKit."""

    __test__ = False  # Prevent pytest from collecting this class

    schema_id: str
    """The schema ID to validate against (required)."""

    schema_version: Optional[str] = None
    """Optional schema version to validate against."""

    server_address: str = "localhost:50051"
    """Server address (default: "localhost:50051")."""

    api_key: Optional[str] = None
    """API key for authentication (optional)."""

    tls: Optional[TLSOptions] = None
    """TLS configuration for secure connections."""


# Backwards compatibility alias
TestConfig = ProducerTestConfig


@dataclass
class MockResponse:
    """Response data for validation."""

    __test__ = False  # Prevent pytest from collecting this class

    status_code: int
    """HTTP status code."""

    body: Any = None
    """Response body (can be any JSON-serializable type)."""

    headers: Dict[str, str] = field(default_factory=dict)
    """Response headers."""


# Backwards compatibility alias
TestResponseData = MockResponse


@dataclass
class MockRequest:
    """Request context for path parameter extraction (optional)."""

    __test__ = False  # Prevent pytest from collecting this class

    method: str = ""
    """HTTP method."""

    path: str = ""
    """Request path."""

    body: Any = None
    """Request body (can be any JSON-serializable type)."""

    headers: Dict[str, str] = field(default_factory=dict)
    """Request headers."""


# Backwards compatibility alias
TestRequestContext = MockRequest


@dataclass
class ProducerValidationResult:
    """Validation result from producer testing."""

    __test__ = False  # Prevent pytest from collecting this class

    valid: bool
    """Whether the validation passed."""

    errors: List[str]
    """Validation error messages (empty if valid)."""

    validated_against_version: Optional[str] = None
    """Schema version used for validation."""

    validated_against_hash: Optional[str] = None
    """Schema hash used for validation."""


# Backwards compatibility alias
TestValidationResult = ProducerValidationResult


class ProducerTestKit:
    """
    Enables schema compliance testing for producers.

    Allows producers to validate their API responses against their OpenAPI
    schema without needing real consumers. Use it in your test suite to ensure
    your handler output matches your contract.

    Example:
        >>> import pytest
        >>> from cvt_sdk.producer import ProducerTestKit, ProducerTestConfig
        >>>
        >>> @pytest.fixture
        ... def test_kit():
        ...     kit = ProducerTestKit(ProducerTestConfig(
        ...         schema_id="user-api",
        ...         server_address="localhost:50051",
        ...     ))
        ...     yield kit
        ...     kit.close()
        >>>
        >>> def test_get_user(test_kit):
        ...     # Call your handler
        ...     response = my_handler.get_user("123")
        ...
        ...     # Validate against schema
        ...     result = test_kit.validate_response(
        ...         method="GET",
        ...         path="/users/123",
        ...         status_code=200,
        ...         body=response,
        ...     )
        ...
        ...     assert result.valid, f"Errors: {result.errors}"
    """

    def __init__(self, config: ProducerTestConfig):
        """
        Creates a new ProducerTestKit instance.

        Args:
            config: Configuration for the test kit.

        Raises:
            ValueError: If schema_id is not provided.
        """
        if not config.schema_id:
            raise ValueError("schema_id is required")

        self._schema_id = config.schema_id
        self._schema_version = config.schema_version
        self._api_key = config.api_key

        if config.tls and config.tls.enabled:
            credentials = self._create_tls_credentials(config.tls)
            self._channel = grpc.secure_channel(config.server_address, credentials)
        else:
            self._channel = grpc.insecure_channel(config.server_address)
        self._client = ContractValidatorStub(self._channel)

    @staticmethod
    def _create_tls_credentials(tls_options: TLSOptions) -> grpc.ChannelCredentials:
        """Creates TLS credentials from the provided options."""
        root_cert: Optional[bytes] = None
        certificate_chain: Optional[bytes] = None
        private_key: Optional[bytes] = None

        if tls_options.root_cert_path:
            root_cert = Path(tls_options.root_cert_path).read_bytes()

        if tls_options.cert_path and tls_options.key_path:
            certificate_chain = Path(tls_options.cert_path).read_bytes()
            private_key = Path(tls_options.key_path).read_bytes()

        return grpc.ssl_channel_credentials(
            root_certificates=root_cert,
            private_key=private_key,
            certificate_chain=certificate_chain,
        )

    def _create_metadata(self) -> list[tuple[str, str]]:
        """Creates gRPC metadata with the API key if configured."""
        metadata = []
        if self._api_key:
            metadata.append(("x-api-key", self._api_key))
        return metadata

    @staticmethod
    def _serialize_body(body: Any) -> str:
        """Converts a body value to a JSON string."""
        if body is None:
            return ""
        if isinstance(body, str):
            return body
        return json.dumps(body)

    def validate_response(
        self,
        method: str,
        path: str,
        status_code: int,
        body: Any = None,
        headers: Optional[Dict[str, str]] = None,
        request: Optional[MockRequest] = None,
    ) -> ProducerValidationResult:
        """
        Validates a producer's response against the registered schema.

        Args:
            method: HTTP method (GET, POST, etc.).
            path: API path with actual values (e.g., /users/123).
            status_code: HTTP status code.
            body: Response body (any JSON-serializable value).
            headers: Response headers (optional).
            request: Optional request context for path parameter extraction.

        Returns:
            A ProducerValidationResult with valid=True if validation succeeds,
            or valid=False with error details if validation fails.

        Example:
            >>> result = test_kit.validate_response(
            ...     method="GET",
            ...     path="/users/123",
            ...     status_code=200,
            ...     headers={"Content-Type": "application/json"},
            ...     body={"id": "123", "name": "John Doe", "email": "john@example.com"},
            ... )
            >>>
            >>> if not result.valid:
            ...     print(f"Validation errors: {result.errors}")
        """
        # Build response data
        response_data = ResponseData(
            status_code=status_code,
            headers=headers or {},
            body=self._serialize_body(body),
        )

        # Build gRPC request
        req = ValidateProducerRequest(
            schema_id=self._schema_id,
            schema_version=self._schema_version or "",
            method=method.upper(),
            path=path,
            response=response_data,
        )

        # Add optional request context if provided
        if request:
            req.request.CopyFrom(
                RequestData(
                    method=request.method or method,
                    path=request.path or path,
                    headers=request.headers or {},
                    body=self._serialize_body(request.body),
                )
            )

        # Call server
        result: ProtoValidationResult = self._client.ValidateProducerResponse(
            req, metadata=self._create_metadata()
        )

        return ProducerValidationResult(
            valid=result.valid,
            errors=list(result.errors),
            validated_against_version=result.validated_against_version or None,
            validated_against_hash=result.validated_against_hash or None,
        )

    def validate_response_data(
        self,
        method: str,
        path: str,
        response: MockResponse,
        request: Optional[MockRequest] = None,
    ) -> ProducerValidationResult:
        """
        Validates a producer's response using a MockResponse object.

        Args:
            method: HTTP method (GET, POST, etc.).
            path: API path with actual values (e.g., /users/123).
            response: Response data to validate.
            request: Optional request context for path parameter extraction.

        Returns:
            A ProducerValidationResult.
        """
        return self.validate_response(
            method=method,
            path=path,
            status_code=response.status_code,
            body=response.body,
            headers=response.headers,
            request=request,
        )

    def validate_interaction(
        self,
        request: MockRequest,
        response: MockResponse,
    ) -> ProducerValidationResult:
        """
        Validates a full interaction (request + response) against the schema.

        This is useful when you want to validate both the request and response
        together, ensuring the complete contract is honored.

        Args:
            request: Request context with method, path, and body.
            response: Response data to validate.

        Returns:
            A ProducerValidationResult.

        Example:
            >>> result = test_kit.validate_interaction(
            ...     request=MockRequest(
            ...         method="POST",
            ...         path="/users",
            ...         body={"name": "John Doe", "email": "john@example.com"},
            ...     ),
            ...     response=MockResponse(
            ...         status_code=201,
            ...         body={"id": "123", "name": "John Doe", "email": "john@example.com"},
            ...     ),
            ... )
        """
        return self.validate_response(
            method=request.method,
            path=request.path,
            status_code=response.status_code,
            body=response.body,
            headers=response.headers,
            request=request,
        )

    def for_endpoint(self, method: str, path_pattern: str) -> "EndpointTester":
        """
        Creates a helper for testing a specific endpoint.

        This is useful when testing multiple scenarios for the same endpoint.

        Args:
            method: HTTP method (GET, POST, etc.).
            path_pattern: API path pattern (e.g., /users/{id}).

        Returns:
            An EndpointTester with a simplified validation interface.

        Example:
            >>> get_user_endpoint = test_kit.for_endpoint("GET", "/users/{id}")
            >>>
            >>> # Test valid response
            >>> result1 = get_user_endpoint.validate_response(
            ...     status_code=200,
            ...     body={"id": "123", "name": "John"},
            ...     path_values={"id": "123"},
            ... )
            >>>
            >>> # Test not found
            >>> result2 = get_user_endpoint.validate_response(
            ...     status_code=404,
            ...     body={"error": "User not found"},
            ...     path_values={"id": "999"},
            ... )
        """
        return EndpointTester(self, method, path_pattern)

    def close(self) -> None:
        """
        Closes the gRPC connection.
        Should be called when the test kit is no longer needed.
        """
        if self._channel:
            self._channel.close()


class EndpointTester:
    """Helper for testing a specific endpoint."""

    def __init__(self, test_kit: ProducerTestKit, method: str, path_pattern: str):
        """Creates an EndpointTester."""
        self._test_kit = test_kit
        self._method = method
        self._path_pattern = path_pattern

    def validate_response(
        self,
        status_code: int,
        body: Any = None,
        headers: Optional[Dict[str, str]] = None,
        path_values: Optional[Dict[str, str]] = None,
    ) -> ProducerValidationResult:
        """
        Validates a response for this endpoint.

        Args:
            status_code: HTTP status code.
            body: Response body (any JSON-serializable value).
            headers: Response headers (optional).
            path_values: Values to substitute in path parameters (optional).

        Returns:
            A ProducerValidationResult.
        """
        actual_path = self._path_pattern

        # Substitute path parameters
        if path_values:
            for key, value in path_values.items():
                actual_path = actual_path.replace(f"{{{key}}}", value)

        return self._test_kit.validate_response(
            method=self._method,
            path=actual_path,
            status_code=status_code,
            body=body,
            headers=headers,
        )

File: sdks/node/src/producer/testing.ts
import * as fs from "fs";
import * as grpc from "@grpc/grpc-js";
import * as protoLoader from "@grpc/proto-loader";
import * as path from "path";

// Resolve proto path relative to the package root
// At runtime: dist/src/producer/testing.js -> ../../../proto/cvt.proto
const PROTO_PATH = path.resolve(
  __dirname,
  "..",
  "..",
  "..",
  "proto",
  "cvt.proto",
);

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});

const protoDescriptor = grpc.loadPackageDefinition(packageDefinition) as any;
const cvt = protoDescriptor.cvt;

/**
 * TLS configuration options for secure connections.
 */
export interface TLSOptions {
  /** Enable TLS for the connection. */
  enabled: boolean;
  /** Path to the root CA certificate file for server verification. */
  rootCertPath?: string;
  /** Path to the client certificate file (for mTLS). */
  certPath?: string;
  /** Path to the client private key file (for mTLS). */
  keyPath?: string;
}

/**
 * Configuration for the ProducerTestKit.
 */
export interface ProducerTestConfig {
  /** The schema ID to validate against (required). */
  schemaId: string;
  /** Optional schema version to validate against. */
  schemaVersion?: string;
  /** Server address (default: "localhost:50051"). */
  serverAddress?: string;
  /** API key for authentication (optional). */
  apiKey?: string;
  /** TLS configuration for secure connections. */
  tls?: TLSOptions;
}

/**
 * Response data for validation.
 */
export interface ResponseData {
  /** HTTP status code. */
  statusCode: number;
  /** Response headers. */
  headers?: Record<string, string>;
  /** Response body (object or string). */
  body?: any;
}

/**
 * Request context for path parameter extraction (optional).
 */
export interface RequestContext {
  /** HTTP method. */
  method?: string;
  /** Request path. */
  path?: string;
  /** Request headers. */
  headers?: Record<string, string>;
  /** Request body. */
  body?: any;
}

/**
 * Validation result from producer testing.
 */
export interface ProducerValidationResult {
  /** Whether the validation passed. */
  valid: boolean;
  /** Validation error messages (empty if valid). */
  errors: string[];
  /** Schema version used for validation. */
  validatedAgainstVersion?: string;
  /** Schema hash used for validation. */
  validatedAgainstHash?: string;
}

/**
 * Parameters for validating a producer response.
 */
export interface ValidateResponseParams {
  /** HTTP method (GET, POST, etc.). */
  method: string;
  /** API path with actual values (e.g., /users/123). */
  path: string;
  /** Response data to validate. */
  response: ResponseData;
  /** Optional request context for path parameter extraction. */
  request?: RequestContext;
}

/**
 * ProducerTestKit enables schema compliance testing for producers.
 *
 * This allows producers to validate their API responses against their OpenAPI
 * schema without needing real consumers. Use it in your test suite to ensure
 * your handler output matches your contract.
 *
 * @example
 * ```typescript
 * import { ProducerTestKit } from '@cvt/node-sdk/producer';
 *
 * describe('User API', () => {
 *   let testKit: ProducerTestKit;
 *
 *   beforeAll(async () => {
 *     // With TLS and API key
 *     testKit = new ProducerTestKit({
 *       schemaId: 'user-api',
 *       serverAddress: 'localhost:50051',
 *       tls: { enabled: true, rootCertPath: './certs/ca.crt' },
 *       apiKey: 'cvt-dev-key-12345',
 *     });
 *   });
 *
 *   afterAll(() => {
 *     testKit.close();
 *   });
 *
 *   it('should return valid user response', async () => {
 *     // Call your handler
 *     const response = await myHandler.getUser('123');
 *
 *     // Validate against schema
 *     const result = await testKit.validateResponse({
 *       method: 'GET',
 *       path: '/users/123',
 *       response: {
 *         statusCode: 200,
 *         body: response,
 *       },
 *     });
 *
 *     expect(result.valid).toBe(true);
 *   });
 * });
 * ```
 */
export class ProducerTestKit {
  private client: any;
  private schemaId: string;
  private schemaVersion?: string;
  private apiKey?: string;

  /**
   * Creates a new ProducerTestKit instance.
   * @param config - Configuration for the test kit.
   */
  constructor(config: ProducerTestConfig) {
    if (!config.schemaId) {
      throw new Error("schemaId is required");
    }

    this.schemaId = config.schemaId;
    this.schemaVersion = config.schemaVersion;
    this.apiKey = config.apiKey;

    const address = config.serverAddress || "localhost:50051";
    let credentials: grpc.ChannelCredentials;
    if (config.tls?.enabled) {
      credentials = this.createTLSCredentials(config.tls);
    } else {
      credentials = grpc.credentials.createInsecure();
    }
    this.client = new cvt.ContractValidator(address, credentials);
  }

  /**
   * Creates TLS credentials from the provided options.
   */
  private createTLSCredentials(
    tlsOptions: TLSOptions,
  ): grpc.ChannelCredentials {
    let rootCert: Buffer | null = null;
    let clientCert: Buffer | undefined;
    let clientKey: Buffer | undefined;

    if (tlsOptions.rootCertPath) {
      rootCert = fs.readFileSync(tlsOptions.rootCertPath);
    }

    if (tlsOptions.certPath && tlsOptions.keyPath) {
      clientCert = fs.readFileSync(tlsOptions.certPath);
      clientKey = fs.readFileSync(tlsOptions.keyPath);
    }

    return grpc.credentials.createSsl(rootCert, clientKey, clientCert);
  }

  /**
   * Creates gRPC metadata with the API key if configured.
   */
  private createMetadata(): grpc.Metadata {
    const metadata = new grpc.Metadata();
    if (this.apiKey) {
      metadata.set("x-api-key", this.apiKey);
    }
    return metadata;
  }

  /**
   * Serializes a body value to a JSON string.
   */
  private serializeBody(body: any): string {
    if (body === undefined || body === null) {
      return "";
    }
    if (typeof body === "string") {
      return body;
    }
    return JSON.stringify(body);
  }

  /**
   * Validates a producer's response against the registered schema.
   *
   * @param params - Validation parameters including method, path, and response data.
   * @returns A promise that resolves with the validation result.
   *
   * @example
   * ```typescript
   * const result = await testKit.validateResponse({
   *   method: 'GET',
   *   path: '/users/123',
   *   response: {
   *     statusCode: 200,
   *     headers: { 'Content-Type': 'application/json' },
   *     body: { id: '123', name: 'John Doe', email: 'john@example.com' },
   *   },
   * });
   *
   * if (!result.valid) {
   *   console.error('Validation errors:', result.errors);
   * }
   * ```
   */
  async validateResponse(
    params: ValidateResponseParams,
  ): Promise<ProducerValidationResult> {
    const { method, path, response, request } = params;

    // Build the gRPC request
    const grpcRequest: any = {
      schema_id: this.schemaId,
      schema_version: this.schemaVersion || "",
      method: method.toUpperCase(),
      path,
      response: {
        status_code: response.statusCode,
        headers: response.headers || {},
        body: this.serializeBody(response.body),
      },
    };

    // Add optional request context if provided
    if (request) {
      grpcRequest.request = {
        method: request.method || method,
        path: request.path || path,
        headers: request.headers || {},
        body: this.serializeBody(request.body),
      };
    }

    return new Promise((resolve, reject) => {
      this.client.ValidateProducerResponse(
        grpcRequest,
        this.createMetadata(),
        (err: any, result: any) => {
          if (err) {
            reject(err);
          } else {
            resolve({
              valid: result.valid,
              errors: result.errors || [],
              validatedAgainstVersion:
                result.validated_against_version || undefined,
              validatedAgainstHash: result.validated_against_hash || undefined,
            });
          }
        },
      );
    });
  }

  /**
   * Validates a full interaction (request + response) against the schema.
   * This is useful when you want to validate both the request and response
   * together, ensuring the complete contract is honored.
   *
   * @param params - Parameters including request and response data.
   * @returns A promise that resolves with the validation result.
   *
   * @example
   * ```typescript
   * const result = await testKit.validateInteraction({
   *   request: {
   *     method: 'POST',
   *     path: '/users',
   *     body: { name: 'John Doe', email: 'john@example.com' },
   *   },
   *   response: {
   *     statusCode: 201,
   *     body: { id: '123', name: 'John Doe', email: 'john@example.com' },
   *   },
   * });
   * ```
   */
  async validateInteraction(params: {
    request: {
      method: string;
      path: string;
      headers?: Record<string, string>;
      body?: any;
    };
    response: ResponseData;
  }): Promise<ProducerValidationResult> {
    return this.validateResponse({
      method: params.request.method,
      path: params.request.path,
      response: params.response,
      request: params.request,
    });
  }

  /**
   * Creates a helper for testing a specific endpoint.
   * This is useful when testing multiple scenarios for the same endpoint.
   *
   * @param method - HTTP method (GET, POST, etc.).
   * @param path - API path pattern (e.g., /users/{id}).
   * @returns An endpoint tester with a simplified validation interface.
   *
   * @example
   * ```typescript
   * const getUserEndpoint = testKit.forEndpoint('GET', '/users/{id}');
   *
   * // Test valid response
   * const result1 = await getUserEndpoint.validateResponse({
   *   statusCode: 200,
   *   body: { id: '123', name: 'John' },
   * }, { pathValues: { id: '123' } });
   *
   * // Test not found
   * const result2 = await getUserEndpoint.validateResponse({
   *   statusCode: 404,
   *   body: { error: 'User not found' },
   * }, { pathValues: { id: '999' } });
   * ```
   */
  forEndpoint(method: string, pathPattern: string) {
    return {
      /**
       * Validates a response for this endpoint.
       * @param response - Response data to validate.
       * @param options - Optional path values to substitute.
       */
      validateResponse: async (
        response: ResponseData,
        options?: { pathValues?: Record<string, string> },
      ): Promise<ProducerValidationResult> => {
        let actualPath = pathPattern;

        // Substitute path parameters if provided
        if (options?.pathValues) {
          for (const [key, value] of Object.entries(options.pathValues)) {
            actualPath = actualPath.replace(`{${key}}`, value);
          }
        }

        return this.validateResponse({
          method,
          path: actualPath,
          response,
        });
      },
    };
  }

  /**
   * Closes the gRPC client connection.
   * Should be called when the test kit is no longer needed.
   */
  close(): void {
    if (this.client) {
      this.client.close();
    }
  }
}

/**
 * Creates a ProducerTestKit with the given configuration.
 * This is an alternative to using `new ProducerTestKit(config)`.
 *
 * @param config - Configuration for the test kit.
 * @returns A new ProducerTestKit instance.
 *
 * @example
 * ```typescript
 * import { createProducerTestKit } from '@cvt/node-sdk/producer';
 *
 * const testKit = createProducerTestKit({
 *   schemaId: 'my-api',
 *   serverAddress: 'localhost:50051',
 * });
 * ```
 */
export function createProducerTestKit(
  config: ProducerTestConfig,
): ProducerTestKit {
  return new ProducerTestKit(config);
}
